hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java

```diff
@@ -398,7 +398,17 @@ public void onFailure(Throwable t) {
         }));
 
         Function<ExecutionPlan, Operation> operationCtor = plan -> new StartExecutionOperation(jobId, executionId);
-        invoke(operationCtor, this::onExecuteStepCompleted, callback);
+        Consumer<Map<MemberInfo, Object>> completionCallback = results -> {
+            executionCancellationCallback = null;
+            onExecuteStepCompleted(results);
```

Comment:
I can see that the job is only restarted when it's TopologyChangedException and auto-restart is enabled, otherwise a restart will not be scheduled.